### DISCUSSION TITLE
how to improve the documentation?

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#1227] [General]
Streamlit is too newbie-friendly and easy to get started, but it has too few customizable parameters per component to have. It's a little troublesome to change it, and although there are many components that can be expanded, it's still not very convenient. NiceGui is also easy to get started, but novices who want to make a beautiful picture like streamlit need to learn a little about various parameters,such as props and classes. and streamlit makes various parameters that need to be configured beautiful by default.        The document is  a little poor in nicegui. 
Just learn these a few days.

**Comment by rodja** [comment]
> The document is a little poor in nicegui.

Do you have suggestions on how to improve?

**Reply by v479038280** [reply]
Here's an example that I happened to see this afternoon（https://reflex.dev/）, providing some out-of-the-box components (e.g. Navibar, vStack,) while retaining a convenient interface to CSS.but its coding is not beautiful and not convinient I think.
Start from a basic tutorial which can walk through some basic concepts (maybe nicegui is too easy to use, as simple as not having to introduce three lines of code), and then the element, and then some component This concept, adding some ready-made available components, can also increase users to share third-party components, such as the ready-to-use navigation bar, you don't need to look at many Tailwindcss and try its properties, you can directly use vstack, hstack, container,you could know its effect accroding to its name ,without go through CSS document and try everything and findout the one needed)  

In addition, a detailed document like this gitbook (https://python.swaroopch.com/) would actually be better than the current document

English not good ,hard to express,sorry for that , Some places may not be described accurately

**Reply by v479038280** [reply]
This style of documentation is more formal, and if we can use nicegui to implement such a layout   [docker document](https://docs.docker.com/engine/reference/commandline/cli/?_gl=1*1hnrcti*_ga*MTQ1MDM2ODIwMC4xNjkxNDEzOTg3*_ga_XJWPQMJYHQ*MTY5MTQyMTUyNy4yLjEuMTY5MTQyMjM0MC4zOS4wLjA.)

**Comment by rodja** [comment]
> providing some out-of-the-box components (e.g. Navibar, vStack,) while retaining a convenient interface to CSS

We have `ui.row`, `ui.column` etc. with `.props`, `.classes` and `.style` which in my view is very comparable to what Reflex offers.

> Start from a basic tutorial which can walk through some basic concepts

Yes, some tutorials would be great. In writing and also videos. Maybe we could create a Wiki here on GitHub as a space to gather and improve tutorials?

> adding some ready-made available components, can also increase users to share third-party components

You mean some kind of component store?

> a detailed document like this gitbook (https://python.swaroopch.com/) would actually be better than the current document

It would be great to have a book about NiceGUI. But I do not think it could or should replace the current documentation. If you just want to look up how something is done, to much text will slow you down.

**Comment by rodja** [comment]
We have started a list of tutorials in our wiki: https://github.com/zauberzeug/nicegui/wiki#tutorials. What do you think?

**Reply by v479038280** [reply]
>We have ui.row, ui.column etc. with .props, .classes and .style which in my view is very comparable to what Reflex offers.

These tools are usefull and powerfull, I want you to see  this pyqt repo [PyQt-Fluent-Widgets](https://github.com/zhiyiYo/PyQt-Fluent-Widgets)
, with some basic widgets like button,slider,toggle , and navigation componet like navigation bar, navgition pannel which is enough for finnal use ,also layout componets v_box_layout,flow_layout,expand layout ,which is suitble for most cases ,don't need to dive into things in the picture below. there is a demo in this repo which demostrates all the widgets. `cd examples/gallery
python demo.py
`
![微信截图_20230810115248](https://github.com/zauberzeug/nicegui/assets/25096045/d2a22bd9-3fe0-4312-ab80-9ad92113b958)
these componet is engough for a final app, and only need to dive into these CSS  in a few cases. The default is the suitble for most cases

I think the above I mentioned is the most critical basic, and the other documents are some supplements.

>Yes, some tutorials would be great. In writing and also videos. Maybe we could create a Wiki here on GitHub as a space to gather and improve tutorials?

Yes ,tha would be much better. If we could put all things about nicegui in one place?
 In the tutorial, the new tools used such as docker and ngnix are better to have a basic introduce according to the instructions we use.

>You mean some kind of component store? 
>It would be great to have a book about NiceGUI. But I do not think it could or should replace the current documentation. If you just want to look up how something is done, to much text will slow you down.

Yes,but  I think that kind of things is just a supplement. A good implementation of the componets offered by nicegui team and  an official  formal  document is enough in most cases. 

>We have started a list of tutorials in our wiki: https://github.com/zauberzeug/nicegui/wiki#tutorials. What do you think?

Practicing these examples  is helpful for people who don't have a good basic. As a community provided routines these are very good, if it is to be organized into the official documentation as an official tutorial, I think it is not that suitable, most people who use nicegui are not specialized in the web, may not familar with nginx or other related tools, it would be better if we could introduce the relevant tools by the way.

**Reply by v479038280** [reply]
All these things may be not necessary when I spent more time practicing the basic especially the layout part  .But this process may be a stuck to new users. When they are learning nicegui It would be better that they don't need to jump out nicegui and go to learn other new things by themself like docker and nginx,fastapi, they just need to focus on nicegui.They only need to dive into these things when they want to adjust some specific features

In fact ,I think nicegui is better than streamlit and reflex ,Why they got so much more stars on github, and reflex just started last year

**Reply by v479038280** [reply]
>All these things may be not necessary when I spent more time practicing the basic especially the layout part 

I say this because there are many tricky things like this ,as shown in the picture below in the red rectangle


![微信截图_20230811021115](https://github.com/zauberzeug/nicegui/assets/25096045/b13bc1b9-a562-4e9a-8324-8312bab58b5b)



![微信截图_20230811021044](https://github.com/zauberzeug/nicegui/assets/25096045/7445eb23-343a-4ef2-ab9b-efb51a703acf)



the 'items-center' seems to move the items vertically center according to the tailwindcss doc,but as demonstrated in the video they all move things horizontally center.


![微信截图_20230811021938](https://github.com/zauberzeug/nicegui/assets/25096045/af8cb72c-b680-4695-be4c-39d4dcacadb7)

**Reply by v479038280** [reply]
If we could make demos like this? Just my opinion 

![微信截图_20230811025059](https://github.com/zauberzeug/nicegui/assets/25096045/c29f8d8b-4203-4f25-b4b7-6abc923a01c5)

**Reply by rodja** [reply]
> If we could make demos like this?

Yes we can. Have you seen [our interactive demo](https://nicegui.io/documentation#try_styling_nicegui_elements_)? Looks like a good start in my opinion.

**Reply by v479038280** [reply]
Actually I know the nicegui can make these things . If we could make a layout gallery in nicegui.io implement these  with nicegui

**Reply by rodja** [reply]
I totally agree with you, @Smug246. A few month back I've tried to come up with an [own layout builder mechanism](https://github.com/zauberzeug/nicegui/pull/364) for NiceGUI which would abstract away most of the CSS stuff. That would surely require its own kind of documentation. But while I'm not giving up on the dream of finding a simpler solution than CSS, it is not very high prioritized right now.

**Reply by v479038280** [reply]
I found a tool could test the  layout  Intuitively dynamically ,would you like to make it part of the officiall codes? we can install 
this tool with` pip insatll ex4nicegui`
![微信截图_20230811121603](https://github.com/zauberzeug/nicegui/assets/25096045/e296b57a-92e6-4d80-bb63-9843cafa8beb)

**Reply by rodja** [reply]
Interesting @v479038280. But I did not find any documentation or project page for the tool and have not been able to install it. Do you know the author?

**Reply by v479038280** [reply]
Here is a video explain the   [layouTool](https://www.bilibili.com/video/BV1CP41147YK/?spm_id_from=333.788&vd_source=da23edd587d81c7c44d31711159c4113),  You could download the code [here](https://github.com/v479038280/nicegui/blob/main/layout_tool.rar)

**Reply by rodja** [reply]
Wow @v479038280 that is really impressive. It seems you have build a debugging tool with which you can inspect and alter layout parameters in any existing NiceGUI app. That is so cool. I would really like this kind of functionality available for NiceGUI.

I think there a two possible scenarios: either we integrate it right into the NiceGUI core or you make it available as a stand alone tool. I'm currently leaning towards having it as a stand-alone project because I'm unsure on how big this really is. The layout tool seems manageable in itself but it requires this undocumented ex4nicegui pip package...

**Reply by v479038280** [reply]
Actually I'm not the auther, He will discuss this with  you later.

**Comment by suspiciousRaccoon** [comment]
Would it be possible to include the titles of the "See more" items in the sidebar? Displaying them when you select one would improve discoverability a lot. 
Something like this:
![image](https://github.com/zauberzeug/nicegui/assets/127566947/ab6303a5-1fff-45c3-ac66-fb79fc4de32f)

**Reply by rodja** [reply]
Good idea. Could you create a pull request?

**Reply by suspiciousRaccoon** [reply]
I'd love to! However, after spending a few hours looking into the page code I'm still not sure how the sidebar is made and handled. I'm unsure how the side_menu() and ui.left_drawer get the data to display. Could you provide some pointers?

**Reply by rodja** [reply]
Yea, it's a bit hidden. The items in the side menu are created here:

https://github.com/zauberzeug/nicegui/blob/d9284cabbc4abd399c212ab2213bd3bd57d28d8c/website/documentation_tools.py#L47-L53

But currently there is no access to the subpage structure when creating the menu. So it will nee quite a refactoring...

**Comment by v479038280** [comment]
I think this [site ](https://panel.holoviz.org/how_to/index.html)is a good example for documentation,
And it is  a good template to make the sidebar and the items in the sidebar.
![微信截图_20230822134723](https://github.com/zauberzeug/nicegui/assets/25096045/a6410537-62ba-45ff-8c8a-088b6463ff02)
